text/0432-contextual-helpers.md

```diff
@@ -0,0 +1,876 @@
+
+console.log(FooBar + 1);
+//          ~~~~~~ FooBar, the constant here
+
+someObj.FooBar = FooBar;
+//               ~~~~~~ well, which one is this?
+```
+
+Clearly, this would be unacceptable and is similar to the situation we find
+ourselves in here.
+
+We propose to issue deprecation warnings whenever we detect these conflicts,
+both at build time and at invocation time, while maintaining the same lookup
+precedence for the time being. For example, when invoking a component in the
+content position, if we see that there is also a helper with the same name, it
+should result in a deprecation asking the developer to remain one or the other.
+
+Notably, there is such a conflict in Ember today where `action` is both a
+helper and a modifier. Instead of deprecating one of them, we propose to use an
+internal mechanism to produce a single special value such that it will be
+invokable as either a modifier or a helper context. This is different than
+"namespace" semantics in that there is only one context-independent value in this
+special case, i.e. `(helper "action") === (modifier "action")`.
+
+We also acknowledge that, so long as there are _implicit_ globals, we may never
+be able to truly unify global bindings with local ones, as implicit global
+bindings have a high risk of conflicting with HTML elements. Consider the
+built-in `input` helper, or an in-app `style` helper. If these were implicitly
+turned into global identifiers, they would conflict with the HTML elements with
+the same name:
+
+```hbs
+<input type="text">
+ ~~~~~ now refers to the global `input` identifier?
+
+<style>...</style>
+ ~~~~~ now refers to the global `style` identifier?
+```
+
+While the problem exists for local bindings also, it was already addressed in
+[RFC #311](https://github.com/emberjs/rfcs/pull/311). With local bindings, this
+problem is fairly noticible and understandable since the conflict is introduced
+nearby. The solution is also fairly simple – just rename the local variable to
+avoid the conflict. With proper linting, this could be quite easily avoided
+altogether.
+
+With _implicit_ global bindings, this problem is might more difficult to spot
+and reason about. There is also no quick way out, other than to rename the
+global component, helper or modifer which could be difficult or not an option
+at all for addon authors trying to maintain compatibility. To truly resolve
+this conflict, we would have to eliminate implicit globals, which is out of
+scope for this RFC. This also wouldn't be a problem until all the proposed
+deprecations are implemented and removed, which would be quite some time.
+
+We _speculate_ that when all the of that is said and done, we would have an
+alternative resolution mechanism ("template imports") that does not have this
+problem. Alternatively, we could exclude the angle bracket invocation position
+from being able to "see" implicit global identifiers.
+
+### Local helpers and modifiers
```

Comment:
Based on the design, it seems like this is not limited to helpers and modifiers. Specifically if the design of this RFC is that these things are "just values" (which I think _is_ roughly what is being said) then this section **also** applies to components.